ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -115,8 +127,89 @@ public void start() throws ReplicationException {
         this.scheduler.scheduleAtFixedRate(persistor, replicationConfig.replicationTokenFlushDelaySeconds,
             replicationConfig.replicationTokenFlushIntervalSeconds, TimeUnit.SECONDS);
       }
+      started = true;
     } catch (IOException e) {
       logger.error("IO error while starting replication", e);
+    } finally {
+      startupLatch.countDown();
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * Note that, this method should be thread-safe because multiple threads (from different cluster change handlers) may
+   * concurrently update remote replica infos.
+   */
+  @Override
+  public void onReplicaAddedOrRemoved(List<ReplicaId> addedReplicas, List<ReplicaId> removedReplicas) {
+    // 1. wait for start() to complete
+    try {
+      startupLatch.await();
+    } catch (InterruptedException e) {
+      logger.warn("Waiting for startup is interrupted.");
```

Comment:
Will throw an IllegalStateException here. Note that, it doesn't help catch up missed event because next time this method is invoked, the pass-in parameters have changed. (For replication manager this is not critical, we still allow whole server to serve traffic temporarily. Missed replica event eventually will generate some error logs and then we can restart this node)